### PR TITLE
Make generic server interfaces more flexible

### LINKED ISSUE #201: Support custom instrumentation for generic gRPC servers
@timostamm re: this feature I would like to discuss possible implementation approaches before I get into coding.

Our gRPC servers (that we plan to migrate to `protobuf-ts` generic servers) heavily use two things:

* all methods are instrumented (latency is measured and exposed via Prometheus metrics)
* methods extract tracing context from gRPC metadata and do something with it.

Here's one of existing wrappers for native `grpc` callbacks: [gist](https://gist.github.com/be9/b0cd255ec76937030e8c639cfc39b165).
I want to package this nicely in the new server library that I create using `protobuf-ts`, but `adaptService` cannot inject any additional code to callbacks that it generates and `ServerCallContext` interface is not extendable.

Here's what I have in mind:

* Add callbacks (like, `beforeUnaryCall`, `afterUnaryCall`, etc.) in a form of options object passed to `adaptService`. These callbacks would receive `ServerCallContext` as an argument.
* Additional fields like tracing context can be set by the callback. Method bodies would get an extended interface as argument:
  ```ts
  interface RicherContext extends ServerCallContext {
    tracingContext: TracingContext;
  }

  class CoolService implements ICoolService {
    // ...
    async meth(request: SomeRequest, context: RicherContext): Promise<SomeResponse> {
      // ...
    }
  }
  ```

Are there better implementation options here, what do you think?

### PR DESCRIPTION
**Pull request by be9**
* Generate server interfaces with a type parameter
* Introduce `contextFactory` option to `adaptService`

Fixes #201

### CODE REVIEW

File: packages/plugin/src/protobufts-plugin.ts [hunk 1]
```diff
@@ -28,7 +28,7 @@ import {ServiceClientGeneratorGrpc} from "./code-gen/service-client-generator-gr
 import * as ts from "typescript";
 
 
-export class ProtobuftsPlugin extends PluginBase {
+export class ProtobuftsPlugin extends PluginBase<GeneratedFile> {
```
Comment (be9) [inline packages/plugin/src/protobufts-plugin.ts:31]: not sure what was wrong , but TS 3.9.10 didn't want to compile this

Comment (timostamm) [reply packages/plugin/src/protobufts-plugin.ts:31]: `PluginBase` got a default type recently: https://github.com/timostamm/protobuf-ts/commit/4304289f4ed83f60d9ba5d6b1b55abc48db84d8d#diff-89c13abbe6b7815deac020cfd77b59a89c17f6773cc969d5ff9bce6c4830d0f1 

Can you try `cd packages/plugin-framework && make build`? That should fix it.

Comment (be9) [reply packages/plugin/src/protobufts-plugin.ts]: ah, right 🤦 
pushed the update

**Review by timostamm — APPROVED** [review]
This is great! Apologies for the delay.

**Pull request merged**